lib/vfs/src/lib.rs

```diff
@@ -22,6 +22,17 @@ pub type Result<T> = std::result::Result<T, FsError>;
 #[repr(transparent)]
 pub struct FileDescriptor(usize);
 
+impl FileDescriptor
+{
+    pub fn new(fd: usize) -> FileDescriptor {
+        FileDescriptor(fd)
+    }
+
+    pub fn fd(&self) -> usize {
+        self.0
+    }
+}
+
```

Comment:
FileDescriptor is intentionally opaque for the user. The internal implementations (host, mem), may use or not a numeric fd

Reply:
Yeah - the reason for this is I actually implemented the polling functionality for WASI that was set as unimplemented. Basically the way I did it was to expose it here:

`impl VirtualFile for Fd {
    fn last_accessed(&self) -> u64 {
        0
    }
    fn last_modified(&self) -> u64 {
        0
    }
    fn created_time(&self) -> u64 {
        0
    }
    fn size(&self) -> u64 {
        0
    }
    fn set_len(&mut self, _new_size: wasi_types::__wasi_filesize_t) -> Result<(), WasiFsError> {
        Err(WasiFsError::PermissionDenied)
    }

    fn unlink(&mut self) -> Result<(), WasiFsError> {
        Ok(())
    }

    fn bytes_available(&self) -> Result<usize, WasiFsError> {
        Ok(0)
    }

    fn get_fd(&self) -> Option<FileDescriptor> {
        Some(FileDescriptor::new(self.raw.id))
    }
}`

...which is the tok terminal project I then used the proxy override functionality to put in a poll implementation without breaking the backwards compatibility with the other users of Wasmer..
https://github.com/john-sharratt/ate/blob/master/tokterm/src/wasi.rs#L238

But you are right we should not really expose the internals - I could drop the polling implementation entirely as it was just for compatibility reasons I implemented it but its not really mandatory.

Alternatively if you have another plan to poll happy to help.